File: src/lab4/java/BubbleSort.java
package lab4.java;

public class BubbleSort implements SortAlgorithm {

    private int[] elems;
    private int posLoop1;
    private int posLoop2 = 1;

    public BubbleSort(int[] elems) {
        this.elems = elems;
        this.posLoop1 = elems.length - 1;

    }

    @Override
    public void step(Swap s) {
        if (elems[posLoop2 - 1] > elems[posLoop2]) {

            s.swap(posLoop2, posLoop2 - 1);
            int temp = elems[posLoop2 - 1];
            elems[posLoop2 - 1] = elems[posLoop2];
            elems[posLoop2] = temp;

        }

        if (posLoop2 == posLoop1) {
            posLoop1--;
            posLoop2 = 1;
        } else {
            posLoop2++;
        }
    }

    @Override
    public boolean isSorted() {
        return posLoop1 == 0;
    }

}

File: src/lab4/BubbleSort.kt
package lab4

class BubbleSort(override val elems: Array<Int>) : SortAlgorithm {

    override fun run(doRepaint: (Int) -> Unit) {

        val n: Int = elems.size

        for (i in 0 until n) {

            for (j in 1 until n - i) {

                if (elems[j - 1] > elems[j]) {
                    val temp = elems[j - 1]
                    elems[j - 1] = elems[j]
                    elems[j] = temp
                }
                doRepaint(j)

            }

        }

        // Elimina la linea roja que queda al final.
        doRepaint(-1)

    }

}
